Add conversation docs (#331)
grammyjs/website: site/docs/plugins/conversations.md

# Conversations (`conversations`)

Create powerful conversational interfaces with ease.

## Introduction

Most chats consist of more than just one single message. (duh)

For example, you may want to ask the user a question, and then wait for the response.
This may even go back and forth several times, so that a conversation unfolds.

> When you think about [middleware](/guide/middleware.md), you will notice that everything is based around a single [context object](/guide/context.md) per handler.

Reviewer: Please, use relative links, so will work with languages

## Conversation Builder Functions

First of all, lets import a few things.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
import {
  type Conversation,
  type ConversationFlavor,
  conversations,
  createConversation,
} from "@grammyjs/conversations";
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
const {
  conversations,
  createConversation,
} = require("@grammyjs/conversations");
```

</CodeGroupItem>
 <CodeGroupItem title="Deno">

```ts
import {
  type Conversation,
  type ConversationFlavor,
  conversations,
  createConversation,
} from "https://deno.land/x/grammy_conversations/mod.ts";
```

</CodeGroupItem>
</CodeGroup>

With that out of the way, we can now have a look at how to define conversational interfaces.

The main element of a conversation is a function with two arguments.
We call this the _conversation builder function_.

```js
async function greeting(conversation, ctx) {
  // TODO: code the conversation
}
```

Let's see what the two parameters are.

**The second parameter** is not that interesting, it is just a regular context object.
As always, it is called `ctx` and uses your custom context type (maybe called `MyContext`).

> The conversations plugin exports a [context flavor](/guide/context.html#additive-context-flavors) called `ConversationFlavor`.

Reviewer: Please, use relative links, so will work with languages

**The first parameter** is the central element of this plugin.
It is commonly named `conversation`, and it has the type `Conversation` ([API reference](https://doc.deno.land/https://deno.land/x/grammy_conversations/mod.ts/~/Conversation)).
It can be used as a handle to control the conversation, such as waiting for user input, and more.
The type `Conversation` expects your custom context type as a type parameter, so you would often use `Conversation<MyContext>`.

In summary, in TypeScript, your conversation builder function will look like this.

```ts
type MyContext = Context & ConversationFlavor;
type MyConversation = Conversation<MyContext>;

async function greeting(conversation: MyConversation, ctx: MyContext) {
  // TODO code the conversation
}
```

Inside of your conversation builder function, you can now define how the conversation should look.

> Before we go in depth about every feature of this plugin, let's have a look at a more complex example than [the simple one](#simple-example) above.

Reviewer 2: ```suggestion
Before we go in depth about every feature of this plugin, let's have a look at a more complex example than the [simple one](#simple-example) above.
```

  await ctx.reply("How many favorite movies do you have?");
  const count = await conversation.form.number();
  const movies: string[] = [];
  for (let i = 0; i < count; i++) {
    await ctx.reply(`Tell me number ${i + 1}!`);
    const titleCtx = await conversation.waitFor(":text");
    movies.push(titleCtx.msg.text);
  }
  await ctx.reply("Here is a better ranking!");
  movies.sort();
  await ctx.reply(movies.map((m, i) => `${i + 1}. ${m}`).join("\n"));
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
async function movie(conversation, ctx) {
  await ctx.reply("How many favorite movies do you have?");
  const count = await conversation.form.number();
  const movies = [];
  for (let i = 0; i < count; i++) {
    await ctx.reply(`Tell me number ${i + 1}!`);
    const titleCtx = await conversation.waitFor(":text");
    movies.push(titleCtx.msg.text);
  }
  await ctx.reply("Here is a better ranking!");
  movies.sort();
  await ctx.reply(movies.map((m, i) => `${i + 1}. ${m}`).join("\n"));
}
```

</CodeGroupItem>
</CodeGroup>

Can you figure out how this bot will work?

## Installing and Entering a Conversation

> First of all, you **must** use the [session plugin](/plugins/session.md) if you want to use the conversations plugin.

Reviewer: Please, use relative links, so will work with languages

::: tip Change the Conversation Identifier
By default, you have to pass the name of the function to `ctx.conversation.enter()`.
However, if you prefer to use a different identifier, you can specify it like so:

```ts
bot.use(createConversation(greeting, "new-name"));
```

In turn, you can enter the conversation with it:

```ts
bot.command("start", (ctx) => ctx.conversation.enter("new-name"));
```

:::

## Leaving a Conversation

The conversation will run until your conversation builder function completes.
This means that you can simply leave a conversation by using `return`.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
async function hiAndBye(conversation: MyConversation, ctx: MyContext) {
  await ctx.reply("Hi! And Bye!");
  // Leave the conversation:
  return;
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
async function hiAndBye(conversation, ctx) {
  await ctx.reply("Hi! And Bye!");
  // Leave the conversation:
  return;
}
```

</CodeGroupItem>
</CodeGroup>

(Yes, putting a `return` at the end of the function is a bit pointless, but you get the idea.)

You can also throw an error.
This will likewise exit the conversation.

> Remember to [install an error handler](/guide/errors.md) on your bot.

Reviewer: Please, use relative links, so will work with languages

If you want to hard-kill the conversation while it is waiting for user input, you can also use `await ctx.conversation.exit()`.
It's often better to stick with simply returning from the function, but there are a few examples where using `await ctx.conversation.exit()` is convenient.
Remember that you must `await` the call.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
// Install the conversations plugin.

> bot.use(conversations());

Reviewer 2: This line is missing in the JavaScript.

bot.command("movie", (ctx) => ctx.conversation.enter("movie"));
```

</CodeGroupItem>
</CodeGroup>

Note that the order matters here.
You must first install the conversations plugin (line 2) before you can call `await ctx.conversation.exit()`.
Also, the generic cancel handlers must be installed before the actual conversations are registered.

## Waiting for Updates

You can use the conversation handle `conversation` to wait for the next update in this particular chat.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
async function waitForMe(conversation: MyConversation, ctx: MyContext) {
  // Wait for the next update:
  const newContext = await conversation.wait();
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
async function waitForMe(conversation, ctx) {
  // Wait for the next update:
  const newContext = await conversation.wait();
}
```

</CodeGroupItem>
</CodeGroup>

An update can mean that a text message was sent, or a button was pressed, or something was edited, or virtually any other action was performed by the user.
Check out the full list in the Telegram docs [here](https://core.telegram.org/bots/api#update).

> Usually, outside of the conversations plugin, every one of these updates would be handled by [the middleware system](/guide/middleware.md) of your bot.

Reviewer 2: ```suggestion
Usually, outside of the conversations plugin, each of these updates would be handled by the [middleware system](/guide/middleware.md) of your bot.
```

Check out the full list in the Telegram docs [here](https://core.telegram.org/bots/api#update).

> Usually, outside of the conversations plugin, each of these updates would be handled by the [middleware system](/guide/middleware.md) of your bot.

Reviewer: Please, use relative links, so will work with languages

Hence, your bot would handle the update via a context object which gets passed to your handlers.

In conversations, you will obtain this new context object from the `wait` call.
In turn, you can handle different updates differently based on this object.
For example, you can check for text messages:

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
async function waitForText(conversation: MyConversation, ctx: MyContext) {
  // Wait for the next update:
  ctx = await conversation.wait();
  // Check for text:
  if (ctx.message?.text) {
    // ...
  }
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
async function waitForText(conversation, ctx) {
  // Wait for the next update:
  ctx = await conversation.wait();
  // Check for text:
  if (ctx.message?.text) {
    // ...
  }
}
```

</CodeGroupItem>
</CodeGroup>

In addition, there are a number of other methods alongside `wait` that let you wait for specific updates only.

> One example is `waitFor` which takes a [filter query](/guide/filter-queries.md) and then only waits for updates that match the provided query.

Reviewer: Please, use relative links, so will work with languages

```ts
// BAD
if (Math.random() < 0.5) { /* do stuff */ }
// GOOD
if (conversation.random() < 0.5) { /* do stuff */ }
```

### Rule III: Use Convenience Functions

There are a bunch of things installed on `conversation` which may greatly help you.
Your code doesn't exactly break if you don't use them, but it can be slow or behave in a confusing way.
The end user might not notice a difference, though.

```ts
// Sleep via conversation, massive performance improvement
await conversation.sleep(3000); // 3 seconds

// Debug logging via conversation, does not print confusing logs
conversation.log("Hello, world");
```

Note that you can do all of the above via `conversation.external()`, but this can be tedious to type, so it's just easier to use the convenience functions.

## Variables, Branching, and Loops

If you follow the three rules above, you are completely free to use any code you like.
We will now go through a few concepts that you already know from programming, and show how they translate to clean and readable conversations.

Imagine that all code below is written inside a conversation builder function.

> You can declare variables and do whatever you want with them.

Reviewer 2: ```suggestion
You can declare variables and do whatever you want with them:
```

```ts
await ctx.reply("Send me your favorite numbers, separated by commas!");
const { message } = await conversation.wait();
const sum = message.text
  .split(",")
  .map((n) => parseInt(n.trim(), 10))
  .reduce((x, y) => x + y);
await ctx.reply("The sum of these numbers is: " + sum);
```

> Branching works, too.

Reviewer 2: ```suggestion
Branching works, too:
```

You can now use it in your main conversation builder function like this:

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
async function enterGroup(conversation: MyConversation, ctx: MyContext) {
  const ok = await captcha(conversation, ctx);

  if (ok) await ctx.reply("Welcome!");
  else await ctx.banChatMember();
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
async function enterGroup(conversation, ctx) {
  const ok = await captcha(conversation, ctx);

  if (ok) await ctx.reply("Welcome!");
  else await ctx.banChatMember();
}
```

</CodeGroupItem>
</CodeGroup>

See how the captcha function can be reused in different places in your code.

> This simple example is only meant to illustrate how functions work.
> In reality, it may work poorly because it only waits for a new update from the respective chat, but without verifying that it actually comes from the same user who joined.
> If you want to create a real captcha, you may want to use [parallel conversations](#parallel-conversations).

If you want, you can also split your code across even more fuctions, or use recursion, mutual recursion, generators, and so on.

> (Just make sure that all functions follow [the three rules](#golden-rules-of-conversations).)

Reviewer 2: ```suggestion
(Just make sure that all functions follow the [three rules](#golden-rules-of-conversations).)
```

Naturally, you can use error handling in your functions, too.
Regular `try`/`catch` statements work just fine, also across functions.
After all, conversations are just JavaScript.

> If the main conversation function throws, the error will propagate further into [the error handling mechanisms](/guide/errors.md) of your bot.

Reviewer 2: ```suggestion
If the main conversation function throws, the error will propagate further into the [error handling mechanisms](/guide/errors.md) of your bot.
```

After all, conversations are just JavaScript.

> If the main conversation function throws, the error will propagate further into the [error handling mechanisms](/guide/errors.md) of your bot.

Reviewer: Please, use relative links, so will work with languages

Naturally, you can just move your functions across modules.
That way, you can define some functions in one file, `export` them, and then `import` and use them in another file.

If you want, you can also define classes.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
class Auth {
  public token?: string;

  constructor(private conversation: MyConversation) {}

  authenticate(ctx: MyContext) {
    const link = getAuthLink(); // get auth link from your system
    await ctx.reply(
      "Open this link to obtain a token, and sent it back to me: " + link,
    );
    ctx = await this.conversation.wait();
    this.token = ctx.message?.text;
  }

  isAuthenticated(): this is Auth & { token: string } {
    return this.token !== undefined;
  }
}

async function askForToken(conversation: MyConversation, ctx: MyContext) {
  const auth = new Auth(conversation);
  await auth.authenticate(ctx);
  if (auth.isAuthenticated()) {
    const token = auth.token;
    // do stuff with token
  }
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```ts
class Auth {
  constructor(conversation) {

>     this.conversation = conversation;

Reviewer 2: Maybe `this.conversation` can be `this.#conversation` as it is private that way.

  await auth.authenticate(ctx);
  if (auth.isAuthenticated()) {
    const token = auth.token;
    // do stuff with token
  }
}
```

</CodeGroupItem>
</CodeGroup>

The point here is not so much that we strictly recommend you to do this.
It is rather meant as an example for how you can use the endless flexibilities of JavaScript to structure your code.

## Forms

As mentioned [earlier](#waiting-for-updates), there are several different utility functions on the conversation handle, such as `await conversation.waitFor('message:text')` which only returns text message updates.

If these methods are not enough, the conversations plugin provides even more helper functions for building forms via `conversation.form`.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
async function waitForMe(conversation: MyConversation, ctx: MyContext) {
  await ctx.reply("How old are you?");
  const age: number = await conversation.form.number();
}
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```js
async function waitForMe(conversation, ctx) {
  await ctx.reply("How old are you?");
  const age = await conversation.form.number();
}
```

</CodeGroupItem>
</CodeGroup>

> As always, check out [the API reference of this plugin](https://doc.deno.land/https://deno.land/x/grammy_conversations@v0.6.2/mod.ts/~/ConversationForm) to see which methods are available.

Reviewer 2: ```suggestion
As always, check out [the API reference of this plugin](https://doc.deno.land/https://deno.land/x/grammy_conversations@v0.7.0/form.ts/~/Form) to see which methods are available.
```

```

### Inspecting Active Conversations

You can see how many conversations with which identifier are running.

```ts
const stats = ctx.conversation.active;
console.log(stats); // { "enterGroup": 1 }
```

This will be provided as an object that has the conversation identifiers as keys, and a number indicating the number of running conversations for each identifier.

## How It Works

> [Remember](#three-golden-rules-of-conversations) that the code inside your conversation builder functions must follow three rules.
> We are now going to see _why_ you need to build them that way.

We are first going to see how this plugin works conceptually, before we elaborate on some details.

### How `wait` Calls Work

Let us switch perspectives for a while, and ask a question from a plugin developer's point of view.
How to implement a `wait` call in the plugin?

The naïve approach to implementing a `wait` call in the conversations plugin would be to create a new promise, and to wait until the next context object arrives.
As soon as it does, we resolve the promise, and the conversation can continue.

However, this is a bad idea for several reasons.

**Data Loss.**

> What if your server crashes while waiting for a context object?

Reviewer: ```suggestion

What if your server crashes while waiting for a context object?
```

For me, it's still hard to read. I suggest adding a blank line to make a "subtitle" or make a list

Reviewer 3: Is it still relevant?

In that case, we lose all information about the state of the conversation.
Basically, the bot loses its train of thought, and the user has to start over.
This is bad desgin and annoying.

**Blocking.**
If wait calls would block until the next update arrives, it means that the middleware execution for the first update can't complete until the entire conversation completes.

- For built-in polling, this means that no further updates can be processed until the current one is done.
  Hence, the bot would simply be blocked forever.

> - For [the grammY runner](./runner.md), the bot would not be blocked.

Reviewer 2: ```suggestion
- For [grammY runner](./runner.md), the bot would not be blocked.
```

  However, when processing thousands of conversations in parallel with different users, it would consume potentially very large amounts of memory.
  If many users stop responding, this leaves the bot stuck in the middle of countless conversations.

> - Webhooks have their own whole [category of problems](/guide/deployment-types.html#ending-webhook-requests-in-time) with long-running middleware.

Reviewer: Please, use relative links, so will work with languages

**State.**
On serverless infrastructure such as cloud functions, we cannot actually assume that the same instance handles two subsequent updates from the same user.
Hence, if we were to create stateful conversations, they may randomly break all the time, as some `wait` calls don't resolve, but some other middleware is suddenly executed.
The result is an abundance of random bugs and chaos.

There are more problems, but you get the idea.

Consequently, the conversations plugin does things differently.
Very differently.

The conversations plugin tracks the execution of your function.
When a wait call is reached, it serializes the state of execution into the session, and safely stores it in a database.
When the next update arrives, it first inspects the session data.
If it finds that it left off in the middle of a conversation, it deserializes the state of execution, takes your conversation builder function, and replays it up to the point of the last `wait` call.
It then resumes ordinary execution of your function—until the next `wait` call is reached, and the execution must be haltet again.

What do we mean by the state of execution?
In a nutshell, it consists on three things:

> 1. Incoming updates.
> 2. Outgoing API calls.
> 3. External events and effects, such as randomness or calls to external APIs or databases.

Reviewer 2: ```suggestion
1. Incoming updates
2. Outgoing API calls
3. External events and effects, such as randomness or calls to external APIs or databases
```

What do we mean by replaying?
Replaying simply means calling the function regularly from the start, but when it does things like calling `wait` or performing API calls, we don't actually do any of those.
Instead, we check out or logs where we recorded from a previous run which values were returned.
We then inject these values so that the conversation builder function simply runs through very fast---until our logs are exhausted.
At that point, we switch back to normal execution mode, which is just a fancy way of saying that we stop injecting stuff, and start to actually perform API calls again.

This is why the plugin has to track all incoming updates as well as all Bot API calls.
(See points 1 and 2 above.)
However, the plugin has no control over external events, side-effects, or randomness.
For example, you could this:

```ts
if (Math.random() < 0.5) {
  // do one thing
} else {
  // do another thing
}
```

In that case, when calling the function, it may suddenly behave differently every time, so replaying the function will break!
It could randomly work differently than the original execution.

> This is why point 3 exists, and [The Three Golden Rules](#three-golden-rules-of-conversations) must be followed.

Reviewer 2: ```suggestion
This is why point 3 exists, and the [Three Golden Rules](#three-golden-rules-of-conversations) must be followed.
```

Author: Do you want to see this applied across the entire website?

Reviewer 2: I think it is already like that, at least places where I have modified

Author: I did a quick case-insensitive search for `[the` and it seems like there are dozens of places that include the _the_ into the link.

Reviewer 2: That doesn't mean you can't commit this suggestion for now.

Author: I will rather fix all other places in the same file with the same commit, and open a new PQ that fixes everything else.
